Show correct WooCommerce pages status if they are using blocks vs shortcode.

budzanowski opened the pull request:
This PR updates the WooCommerce status page to also consider WooCommerce Blocks for the Cart and Checkout pages status. Additionally, it displays the shortcode and or block slug in case they were not found ( it looks like this was the original intention some time ago bc the `sprintf` argument was there - just the placeholder was missing.

![image](https://user-images.githubusercontent.com/17271089/103225553-7082b380-492a-11eb-9639-323258fabe4f.png)

In the case when previously a page that had checkout block was used for checkout ( the same situation for cart ) the status page was showing that the page is missing the shortcode. Now the status is correct.

![image](https://user-images.githubusercontent.com/17271089/103225652-b6d81280-492a-11eb-8bd0-04cbd03a4976.png)

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes woocommerce/woocommerce-gutenberg-products-block/issues/3565 .

### How to test the changes in this Pull Request:

#### regression test
1. WC pages set up correctly, Cart and Checkout pages ( shortcode ) exist
2. Go to the WC status page and check that the pages are correctly detected

#### missing setup test
1. Select a page that has no shortcode or block for the Cart ( or Checkout ) page.
2. Go to the status view.
3. There should be an indication that the shortcode or the block is missing.

#### block present
1. Select a page that has the woocommerce/checkout block for the Checkout page.
2. Go to the status view.
3. The Checkout should be marked as green.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Check if Cart and Checkout pages are using block instead of shortcode for the Status check.

budzanowski committed:
Add block information for woocommerce pages report.

budzanowski committed:
Print missing block info in the status page.

Aljullu commented on includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1220,6 +1227,16 @@ public function get_pages() {
+			if ( $values['block'] && get_post( $page_id ) ) {
+				$block_required = true;
+				$page           = get_post( $page_id );
+				$page_info      = WC_Tracker::get_block_tracker_data( $values['block'], $page_name );
```
I feel a bit weird calling `WC_Tracker` for something which is not really related to the tracker.

Out of the scope for this PR, but at some point we might want to abstract that method somewhere outside the tracker.

budzanowski replied:
Absolutely, it feels wired, but the use is so small that I did not want to abstract too early.

haszari replied:
FYI the checks in `get_block_tracker_data` are a bit primitive, could be improved - https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3224#issuecomment-707391821

There's a new improved approach using `parse_blocks()` in a PR that didn't merge, could be adapted into core: https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/bbef0a6763f3c5073902be248820badc09bc2496

It's might be worthwhile spending a bit of time tidying this up and abstracting out the detection logic. We don't often touch this code, so possibly worth investing a little more effort?

budzanowski replied:
I don't mind doing it, cleanup makes sense. But I don't want to do it as a part of this PR. The scope should be separated. I don't mind postponing this and working on a cleaned-up version if this is what we think this is how this should be done.

budzanowski replied:
Or merge as it is and follow up with a cleanup.

nerrad replied:
A couple thoughts from my perspective here:

- I don't like using `WC_Tracker:: get_block_tracker_data` as this on the surface appears to be a violation of separation of concerns. For now the method does what you want, but it is a part of a class dedicated to WC_Tracker events so there's no guarantee it will do what you want in future refactors (or do what you want without any side-effects).
- To that point, extraction of whether blocks are in a page should not even be a method on `WC_Tracker` itself but rather something `WC_Tracker` consumes as a service.

I recognize the desire to keep scope from creeping in this pull and normally I'd agree, but this is severe enough that I'd like to avoid compounding the tech debt that already exists by pulling in `WC_Tracker` usage. I took a brief look at the code Rua shared and it should be trivial enough to create a new service class in Woo for explicitly solving this block detection problem (using the code Rua shared, or some variation of it). If you want to avoid creating an entire new class, you can add it as private methods in this class (`WC_REST_System_Status_V2_Controller`) for now and future follow ups could extract to a service class and implement in WC_Tracker too. I'd prefer creating the class.

Aljullu reviewed (dismissed):
Nice to see this getting fixed! Code looks good to me and it tests well. :ok_hand:

nerrad commented on includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1172,22 +1172,27 @@ public function get_pages() {
 			_x( 'Cart', 'Page setting', 'woocommerce' ) => array(
 				'option'    => 'woocommerce_cart_page_id',
 				'shortcode' => '[' . apply_filters( 'woocommerce_cart_shortcode_tag', 'woocommerce_cart' ) . ']',
+				'block'     => apply_filters( 'woocommerce_cart_block_tag', 'woocommerce/cart' ),
```
Why is this (or for that matter shortcode too... but that's already a done deal) filtered? does it need to be filtered?

budzanowski replied:
This allows 3rd party plugins to provide valid blocks that can act as cart or checkout. If we don't filter it then `woocommerce/cart` becomes the only valid block here.

nerrad replied:
Rationale sounds fair, I just don't think we should expose this filter until someone demonstrates why they'd want it. Not a blocker imo though, but probably should be weighed in on by @woocommerce/proton.

nerrad reviewed (commented):
Coming along great! Just adding my two cents.

budzanowski committed:
Merge branch 'master' into fix/woocommerce-gutenberg-products-block-issues-3565

budzanowski committed:
Refactor.

budzanowski committed:
Ups.

budzanowski commented:
I have refactored the code in a way that the structure should now be more logical and maintainable. I have created a new utils class. I have moved some methods ( private ) from the tracker class there so we now have a central point for blocks related utils. I have also added a small and simple utility function that checks if a page contains a block.
@Aljullu @haszari @nerrad ?

nerrad commented on includes/blocks/class-wc-blocks-utils.php:
```diff
@@ -0,0 +1,84 @@
+		);
+	}
+
+	/**
+	 * Check if a given page contains a particular block.
+	 *
+	 * @param string $page       (int|WP_Post) Page post ID or post object.
+	 * @param string $block_name The name (id) of a block, e.g. `woocommerce/cart`.
+	 * @return bool|null         Boolean value if the page contains the block or not. Null in case the page does not exist.
+	 */
+	public static function does_page_contains_block( $page, $block_name ) {
```
minor nit, grammar - should be `does_page_contain_block`. However, I also think I'd prefer something like this:

```suggestion
	/**
	 * Check if a given page contains a particular block.
	 *
	 * @param int|WP_Post $page   Page post ID or post object.
	 * @param string $block_name The name (id) of a block, e.g. `woocommerce/cart`.
	 * @return bool|null         Boolean value if the page contains the block or not. Null in case the page does not exist.
	 */
	public static function has_block_in_page( $page, $block_name ) {
```

I find following the convention of `has_` or `is_` prefixes for methods that are checks returning boolean to be really great to use for easier recall.

Also, I get why you're returning null here for when a page doesn't exist, but for the purpose of this method, I think it's best to not support that use case and mix concerns. It should only really be used to determine if the block exists in the page or not (if the page doesn't exist that's irrelevant?).

nerrad commented on includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1220,6 +1227,14 @@ public function get_pages() {
+			// Block checks.
+			if ( $values['block'] && get_post( $page_id ) ) {
+				$block_required = true;
+				if ( WC_Blocks_Utils::does_page_contains_block( $page_id, $values['block'] ) ) {
+					$block_present = true;
+				}
```
Can probably just do something like:

```suggestion
				$block_present = WC_Block_Utils:does_page_contains_block( $page_id, $values['block'] );
```

nerrad reviewed (commented):
Looks great Bartek! Thanks for handling the feedback. Probably should add some phpunit tests for the new utility class. Also probably should get the 👌 from someone on @woocommerce/proton before merging.

budzanowski committed:
Update includes/blocks/class-wc-blocks-utils.php

Co-authored-by: Darren Ethier <darren@roughsmootheng.in>

budzanowski committed:
Review refactor.

budzanowski committed:
Simplify.

haszari commented:
Looks good to me, nice work @budzanowski !

budzanowski committed:
Add block utils unit tests.

budzanowski commented:
@nerrad Unit Tests are done. Now I need to find someone from Proton for a review :P

jonathansadowski commented on includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1172,22 +1172,27 @@ public function get_pages() {
 			_x( 'Cart', 'Page setting', 'woocommerce' ) => array(
 				'option'    => 'woocommerce_cart_page_id',
 				'shortcode' => '[' . apply_filters( 'woocommerce_cart_shortcode_tag', 'woocommerce_cart' ) . ']',
+				'block'     => apply_filters( 'woocommerce_cart_block_tag', 'woocommerce/cart' ),
```
Thanks for the explanation @budzanowski, but I agree with @nerrad that we should hold off on this filter. Can we remove the filter before moving forward with this?

budzanowski replied:
Done.

jonathansadowski commented on includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1172,22 +1172,27 @@ public function get_pages() {
 			_x( 'Checkout', 'Page setting', 'woocommerce' ) => array(
 				'option'    => 'woocommerce_checkout_page_id',
 				'shortcode' => '[' . apply_filters( 'woocommerce_checkout_shortcode_tag', 'woocommerce_checkout' ) . ']',
+				'block'      => apply_filters( 'woocommerce_checkout_block_tag', 'woocommerce/checkout' ),
```
Same rationale to this filter. Thanks.

budzanowski replied:
Done.

jonathansadowski reviewed (dismissed):
Thanks for tackling this. Looking good to me. I agreed with the previous comment about the additional filters, but I think if we address that, this should be good to go.

budzanowski committed:
Remove filters that we don't need yet.

budzanowski commented:
@jonathansadowski review suggestions applied.

claudiosanches reviewed (approved):
Thanks!
